tests/test_dataset.py

```diff
@@ -157,6 +157,32 @@ def test_can_detect(self):
 
             self.assertEqual(DEFAULT_FORMAT, detected_format)
 
+    @mark_requirement(Requirements.DATUM_GENERAL_REQ)
+    def test_can_detect_with_nested_folder(self):
+        env = Environment()
+        env.importers.items = {DEFAULT_FORMAT: env.importers[DEFAULT_FORMAT]}
+        env.extractors.items = {DEFAULT_FORMAT: env.extractors[DEFAULT_FORMAT]}
```

Comment:
The test is generally fine, but consider extending it with a generic "fits everything" format (like `image dir`) to test for collisions and that we continue search till we find one match. Maybe, it should be in another test.

Reply:
It's look that such situation when we have more than one match and continue search till we find one match, will occurs never. Because we continue search if our path has only one file system object and this object it's a directory, but it for this path we have more than one match it's mean that this path most likely contains more than one file system object, so that these two situations are contradictory.

Reply:
For example, we could detect "image_dir" and "imagenet" in any root image directory, but only if we go deeper, we we'll detect a COCO dataset, for example. I know it is implemented already, just asking for a test that ensures this logic works.

Example:
```
a/annotations/x_instances.json
a/images/y.jpg
^
here we detect few general matching formats (currently, imagedir, imagenet, celeba, align_celeba, cifar etc.)
```
And inside `a` we detect COCO.

Reply:
Added this test.